File: apps/web_companion/src/protocol/msg_map.ts
import * as Auth from '@chirp/proto/auth';
import * as Chat from '@chirp/proto/chat';
import * as Party from '@chirp/proto/party';
import * as Social from '@chirp/proto/social';
import { MsgID } from '@chirp/proto/gateway';

/**
 * A typed request/response pair. On this protocol a response is correlated by
 * Packet.sequence, not by msgId — the server answers with the RESP msgId and
 * echoes the request's sequence. The spec keeps both ids and the codec in one
 * place so call sites never touch raw msgIds.
 */
export interface MessageSpec<Req, Resp> {
  readonly reqMsgId: MsgID;
  readonly respMsgId: MsgID;
  /** Accepts a partial; unset fields fall back to proto defaults via fromPartial. */
  encodeRequest(message: Partial<Req>): Uint8Array;
  decodeResponse(bytes: Uint8Array): Resp;
}

interface ProtoMessage<T> {
  encode(message: T, writer?: unknown): { finish(): Uint8Array };
  decode(input: Uint8Array, length?: number): T;
  fromPartial(partial: Partial<T>): T;
}

function defineSpec<Req, Resp>(
  reqMsgId: MsgID,
  respMsgId: MsgID,
  Req: ProtoMessage<Req>,
  Resp: ProtoMessage<Resp>,
): MessageSpec<Req, Resp> {
  return {
    reqMsgId,
    respMsgId,
    encodeRequest: (message) => Req.encode(Req.fromPartial(message)).finish(),
    decodeResponse: (bytes) => Resp.decode(bytes),
  };
}

// Core pair used by the connection layer itself.
export const LOGIN = defineSpec(
  MsgID.LOGIN_REQ,
  MsgID.LOGIN_RESP,
  Auth.LoginRequest,
  Auth.LoginResponse,
);
export const LOGOUT = defineSpec(
  MsgID.LOGOUT_REQ,
  MsgID.LOGOUT_RESP,
  Auth.LogoutRequest,
  Auth.LogoutResponse,
);

// Chat (direct chat WS entry; the full table grows in the api layer).
export const SEND_MESSAGE = defineSpec(
  MsgID.SEND_MESSAGE_REQ,
  MsgID.SEND_MESSAGE_RESP,
  Chat.SendMessageRequest,
  Chat.SendMessageResponse,
);
export const GET_HISTORY = defineSpec(
  MsgID.GET_HISTORY_REQ,
  MsgID.GET_HISTORY_RESP,
  Chat.GetHistoryRequest,
  Chat.GetHistoryResponse,
);
export const MARK_READ = defineSpec(
  MsgID.MARK_READ_REQ,
  MsgID.MARK_READ_RESP,
  Chat.MarkReadRequest,
  Chat.MarkReadResponse,
);
export const GET_UNREAD_COUNT = defineSpec(
  MsgID.GET_UNREAD_COUNT_REQ,
  MsgID.GET_UNREAD_COUNT_RESP,
  Chat.GetUnreadCountRequest,
  Chat.GetUnreadCountResponse,
);
export const GET_USER_GROUPS = defineSpec(
  MsgID.GET_USER_GROUPS_REQ,
  MsgID.GET_USER_GROUPS_RESP,
  Chat.GetUserGroupsRequest,
  Chat.GetUserGroupsResponse,
);
export const CREATE_GROUP = defineSpec(
  MsgID.CREATE_GROUP_REQ,
  MsgID.CREATE_GROUP_RESP,
  Chat.CreateGroupRequest,
  Chat.CreateGroupResponse,
);
export const INVITE_TO_GROUP = defineSpec(
  MsgID.INVITE_TO_GROUP_REQ,
  MsgID.INVITE_TO_GROUP_RESP,
  Chat.InviteToGroupRequest,
  Chat.InviteToGroupResponse,
);
export const LEAVE_GROUP = defineSpec(
  MsgID.LEAVE_GROUP_REQ,
  MsgID.LEAVE_GROUP_RESP,
  Chat.LeaveGroupRequest,
  Chat.LeaveGroupResponse,
);
export const KICK_MEMBER = defineSpec(
  MsgID.KICK_MEMBER_REQ,
  MsgID.KICK_MEMBER_RESP,
  Chat.KickMemberRequest,
  Chat.KickMemberResponse,
);
export const GET_GROUP_MEMBERS = defineSpec(
  MsgID.GET_GROUP_MEMBERS_REQ,
  MsgID.GET_GROUP_MEMBERS_RESP,
  Chat.GetGroupMembersRequest,
  Chat.GetGroupMembersResponse,
);

// Reactions, edit/delete (typing has no REQ: it is fire-and-forget 2208).
export const ADD_REACTION = defineSpec(
  MsgID.ADD_REACTION_REQ,
  MsgID.ADD_REACTION_RESP,
  Chat.AddReactionRequest,
  Chat.AddReactionResponse,
);
export const REMOVE_REACTION = defineSpec(
  MsgID.REMOVE_REACTION_REQ,
  MsgID.REMOVE_REACTION_RESP,
  Chat.RemoveReactionRequest,
  Chat.RemoveReactionResponse,
);
export const EDIT_MESSAGE = defineSpec(
  MsgID.EDIT_MESSAGE_REQ,
  MsgID.EDIT_MESSAGE_RESP,
  Chat.EditMessageRequest,
  Chat.EditMessageResponse,
);
export const DELETE_MESSAGE = defineSpec(
  MsgID.DELETE_MESSAGE_REQ,
  MsgID.DELETE_MESSAGE_RESP,
  Chat.DeleteMessageRequest,
  Chat.DeleteMessageResponse,
);

// Social plane (WS 8001): the backend serves the full roster surface
// (friends, pending requests, blocked list); friends stay pending-request
// based and presence is a broadcast snapshot.
export const ADD_FRIEND = defineSpec(
  MsgID.ADD_FRIEND_REQ,
  MsgID.ADD_FRIEND_RESP,
  Social.AddFriendRequest,
  Social.AddFriendResponse,
);
export const FRIEND_REQUEST_ACTION = defineSpec(
  MsgID.FRIEND_REQUEST_ACTION_REQ,
  MsgID.FRIEND_REQUEST_ACTION_RESP,
  Social.FriendRequestAction,
  Social.FriendRequestActionResponse,
);
export const REMOVE_FRIEND = defineSpec(
  MsgID.REMOVE_FRIEND_REQ,
  MsgID.REMOVE_FRIEND_RESP,
  Social.RemoveFriendRequest,
  Social.RemoveFriendResponse,
);
export const GET_FRIEND_LIST = defineSpec(
  MsgID.GET_FRIEND_LIST_REQ,
  MsgID.GET_FRIEND_LIST_RESP,
  Social.GetFriendListRequest,
  Social.GetFriendListResponse,
);
export const GET_PENDING_REQUESTS = defineSpec(
  MsgID.GET_PENDING_REQUESTS_REQ,
  MsgID.GET_PENDING_REQUESTS_RESP,
  Social.GetPendingRequestsRequest,
  Social.GetPendingRequestsResponse,
);
export const BLOCK_USER = defineSpec(
  MsgID.BLOCK_USER_REQ,
  MsgID.BLOCK_USER_RESP,
  Social.BlockUserRequest,
  Social.BlockUserResponse,
);
export const UNBLOCK_USER = defineSpec(
  MsgID.UNBLOCK_USER_REQ,
  MsgID.UNBLOCK_USER_RESP,
  Social.UnblockUserRequest,
  Social.UnblockUserResponse,
);
export const GET_BLOCKED_LIST = defineSpec(
  MsgID.GET_BLOCKED_LIST_REQ,
  MsgID.GET_BLOCKED_LIST_RESP,
  Social.GetBlockedListRequest,
  Social.GetBlockedListResponse,
);
export const SET_PRESENCE = defineSpec(
  MsgID.SET_PRESENCE_REQ,
  MsgID.SET_PRESENCE_RESP,
  Social.SetPresenceRequest,
  Social.SetPresenceResponse,
);
export const GET_PRESENCE = defineSpec(
  MsgID.GET_PRESENCE_REQ,
  MsgID.GET_PRESENCE_RESP,
  Social.GetPresenceRequest,
  Social.GetPresenceResponse,
);

// Party plane (WS 7501): cross-game team-up. Invite-accept only, snapshot
// sync (PARTY_STATE_CHANGED carries the full PartyInfo to every member).
export const CREATE_PARTY = defineSpec(
  MsgID.CREATE_PARTY_REQ,
  MsgID.CREATE_PARTY_RESP,
  Party.CreatePartyRequest,
  Party.CreatePartyResponse,
);
export const DISBAND_PARTY = defineSpec(
  MsgID.DISBAND_PARTY_REQ,
  MsgID.DISBAND_PARTY_RESP,
  Party.DisbandPartyRequest,
  Party.DisbandPartyResponse,
);
export const INVITE_TO_PARTY = defineSpec(
  MsgID.INVITE_TO_PARTY_REQ,
  MsgID.INVITE_TO_PARTY_RESP,
  Party.InviteToPartyRequest,
  Party.InviteToPartyResponse,
);
export const ACCEPT_PARTY_INVITE = defineSpec(
  MsgID.ACCEPT_INVITE_REQ,
  MsgID.ACCEPT_INVITE_RESP,
  Party.AcceptInviteRequest,
  Party.AcceptInviteResponse,
);
export const DECLINE_PARTY_INVITE = defineSpec(
  MsgID.DECLINE_INVITE_REQ,
  MsgID.DECLINE_INVITE_RESP,
  Party.DeclineInviteRequest,
  Party.DeclineInviteResponse,
);
export const LEAVE_PARTY = defineSpec(
  MsgID.LEAVE_PARTY_REQ,
  MsgID.LEAVE_PARTY_RESP,
  Party.LeavePartyRequest,
  Party.LeavePartyResponse,
);
export const KICK_PARTY_MEMBER = defineSpec(
  MsgID.KICK_PARTY_MEMBER_REQ,
  MsgID.KICK_PARTY_MEMBER_RESP,
  Party.KickMemberRequest,
  Party.KickMemberResponse,
);
export const TRANSFER_PARTY_LEADER = defineSpec(
  MsgID.TRANSFER_LEADER_REQ,
  MsgID.TRANSFER_LEADER_RESP,
  Party.TransferLeaderRequest,
  Party.TransferLeaderResponse,
);
export const SET_PARTY_READY = defineSpec(
  MsgID.SET_READY_REQ,
  MsgID.SET_READY_RESP,
  Party.SetReadyRequest,
  Party.SetReadyResponse,
);
export const GET_MY_PARTY = defineSpec(
  MsgID.GET_MY_PARTY_REQ,
  MsgID.GET_MY_PARTY_RESP,
  Party.GetMyPartyRequest,
  Party.GetMyPartyResponse,
);

File: apps/mobile_companion/lib/protocol/msg_map.dart
// Prefixed proto namespaces for the decoders below; MsgID/ErrorCode and the
// shared enums come from the barrel (bare export).
import 'dart:typed_data';

import 'package:chirp_proto/proto/auth.pb.dart' as auth;
import 'package:chirp_proto/proto/chat.pb.dart' as chat;
import 'package:chirp_proto/proto/notification.pb.dart' as notification;
// party.KickMemberRequest/Response collide with chat's and are hidden from
// the barrel; the party file comes in directly, prefixed.
import 'package:chirp_proto/proto/party.pb.dart' as party;
import 'package:chirp_proto/proto/social.pb.dart' as social;
import 'package:chirp_proto/chirp_proto.dart';
import 'package:protobuf/protobuf.dart';

/// A typed request/response pair. On this protocol a response is correlated
/// by Packet.sequence, not by msgId — the server answers with the RESP msgId
/// and echoes the request's sequence. The spec keeps both ids and the
/// decoder in one place so call sites never touch raw msgIds.
class MessageSpec<TResp extends GeneratedMessage> {
  const MessageSpec(this.reqMsgId, this.respMsgId, this.decodeResponse);

  final MsgID reqMsgId;
  final MsgID respMsgId;

  /// Protoc-plugin constructor tear-off, e.g. `auth.LoginResponse.fromBuffer`
  /// (its List<int> parameter accepts Uint8List arguments).
  final TResp Function(Uint8List bytes) decodeResponse;
}

// Core pair used by the connection layer itself.
const login = MessageSpec<auth.LoginResponse>(
    MsgID.LOGIN_REQ, MsgID.LOGIN_RESP, auth.LoginResponse.fromBuffer);
const logout = MessageSpec<auth.LogoutResponse>(
    MsgID.LOGOUT_REQ, MsgID.LOGOUT_RESP, auth.LogoutResponse.fromBuffer);

// Chat (direct chat WS entry).
const sendMessage = MessageSpec<chat.SendMessageResponse>(
    MsgID.SEND_MESSAGE_REQ,
    MsgID.SEND_MESSAGE_RESP,
    chat.SendMessageResponse.fromBuffer);
const getHistory = MessageSpec<chat.GetHistoryResponse>(MsgID.GET_HISTORY_REQ,
    MsgID.GET_HISTORY_RESP, chat.GetHistoryResponse.fromBuffer);
const markRead = MessageSpec<chat.MarkReadResponse>(MsgID.MARK_READ_REQ,
    MsgID.MARK_READ_RESP, chat.MarkReadResponse.fromBuffer);
const getUserGroups = MessageSpec<chat.GetUserGroupsResponse>(
    MsgID.GET_USER_GROUPS_REQ,
    MsgID.GET_USER_GROUPS_RESP,
    chat.GetUserGroupsResponse.fromBuffer);
const createGroup = MessageSpec<chat.CreateGroupResponse>(
    MsgID.CREATE_GROUP_REQ,
    MsgID.CREATE_GROUP_RESP,
    chat.CreateGroupResponse.fromBuffer);
const inviteToGroup = MessageSpec<chat.InviteToGroupResponse>(
    MsgID.INVITE_TO_GROUP_REQ,
    MsgID.INVITE_TO_GROUP_RESP,
    chat.InviteToGroupResponse.fromBuffer);
const leaveGroup = MessageSpec<chat.LeaveGroupResponse>(MsgID.LEAVE_GROUP_REQ,
    MsgID.LEAVE_GROUP_RESP, chat.LeaveGroupResponse.fromBuffer);
const kickMember = MessageSpec<chat.KickMemberResponse>(MsgID.KICK_MEMBER_REQ,
    MsgID.KICK_MEMBER_RESP, chat.KickMemberResponse.fromBuffer);
const getGroupMembers = MessageSpec<chat.GetGroupMembersResponse>(
    MsgID.GET_GROUP_MEMBERS_REQ,
    MsgID.GET_GROUP_MEMBERS_RESP,
    chat.GetGroupMembersResponse.fromBuffer);

// Reactions, edit/delete (typing has no REQ: it is fire-and-forget 2208).
const addReaction = MessageSpec<chat.AddReactionResponse>(
    MsgID.ADD_REACTION_REQ,
    MsgID.ADD_REACTION_RESP,
    chat.AddReactionResponse.fromBuffer);
const removeReaction = MessageSpec<chat.RemoveReactionResponse>(
    MsgID.REMOVE_REACTION_REQ,
    MsgID.REMOVE_REACTION_RESP,
    chat.RemoveReactionResponse.fromBuffer);
const editMessage = MessageSpec<chat.EditMessageResponse>(
    MsgID.EDIT_MESSAGE_REQ,
    MsgID.EDIT_MESSAGE_RESP,
    chat.EditMessageResponse.fromBuffer);
const deleteMessage = MessageSpec<chat.DeleteMessageResponse>(
    MsgID.DELETE_MESSAGE_REQ,
    MsgID.DELETE_MESSAGE_RESP,
    chat.DeleteMessageResponse.fromBuffer);

// Social plane (WS 8001): friends and presence.
const addFriend = MessageSpec<social.AddFriendResponse>(MsgID.ADD_FRIEND_REQ,
    MsgID.ADD_FRIEND_RESP, social.AddFriendResponse.fromBuffer);
const friendRequestAction = MessageSpec<social.FriendRequestActionResponse>(
    MsgID.FRIEND_REQUEST_ACTION_REQ,
    MsgID.FRIEND_REQUEST_ACTION_RESP,
    social.FriendRequestActionResponse.fromBuffer);
const removeFriend = MessageSpec<social.RemoveFriendResponse>(
    MsgID.REMOVE_FRIEND_REQ,
    MsgID.REMOVE_FRIEND_RESP,
    social.RemoveFriendResponse.fromBuffer);
const getFriendList = MessageSpec<social.GetFriendListResponse>(
    MsgID.GET_FRIEND_LIST_REQ,
    MsgID.GET_FRIEND_LIST_RESP,
    social.GetFriendListResponse.fromBuffer);
const getPendingRequests = MessageSpec<social.GetPendingRequestsResponse>(
    MsgID.GET_PENDING_REQUESTS_REQ,
    MsgID.GET_PENDING_REQUESTS_RESP,
    social.GetPendingRequestsResponse.fromBuffer);
const blockUser = MessageSpec<social.BlockUserResponse>(MsgID.BLOCK_USER_REQ,
    MsgID.BLOCK_USER_RESP, social.BlockUserResponse.fromBuffer);
const unblockUser = MessageSpec<social.UnblockUserResponse>(
    MsgID.UNBLOCK_USER_REQ,
    MsgID.UNBLOCK_USER_RESP,
    social.UnblockUserResponse.fromBuffer);
const getBlockedList = MessageSpec<social.GetBlockedListResponse>(
    MsgID.GET_BLOCKED_LIST_REQ,
    MsgID.GET_BLOCKED_LIST_RESP,
    social.GetBlockedListResponse.fromBuffer);
const setPresence = MessageSpec<social.SetPresenceResponse>(
    MsgID.SET_PRESENCE_REQ,
    MsgID.SET_PRESENCE_RESP,
    social.SetPresenceResponse.fromBuffer);
const getPresence = MessageSpec<social.GetPresenceResponse>(
    MsgID.GET_PRESENCE_REQ,
    MsgID.GET_PRESENCE_RESP,
    social.GetPresenceResponse.fromBuffer);

// Party plane (WS 7501): cross-game team-up. Invite-accept only, snapshot
// sync (PARTY_STATE_CHANGED carries the full PartyInfo to every member).
const createParty = MessageSpec<party.CreatePartyResponse>(
    MsgID.CREATE_PARTY_REQ,
    MsgID.CREATE_PARTY_RESP,
    party.CreatePartyResponse.fromBuffer);
const disbandParty = MessageSpec<party.DisbandPartyResponse>(
    MsgID.DISBAND_PARTY_REQ,
    MsgID.DISBAND_PARTY_RESP,
    party.DisbandPartyResponse.fromBuffer);
const inviteToParty = MessageSpec<party.InviteToPartyResponse>(
    MsgID.INVITE_TO_PARTY_REQ,
    MsgID.INVITE_TO_PARTY_RESP,
    party.InviteToPartyResponse.fromBuffer);
const acceptPartyInvite = MessageSpec<party.AcceptInviteResponse>(
    MsgID.ACCEPT_INVITE_REQ,
    MsgID.ACCEPT_INVITE_RESP,
    party.AcceptInviteResponse.fromBuffer);
const declinePartyInvite = MessageSpec<party.DeclineInviteResponse>(
    MsgID.DECLINE_INVITE_REQ,
    MsgID.DECLINE_INVITE_RESP,
    party.DeclineInviteResponse.fromBuffer);
const leaveParty = MessageSpec<party.LeavePartyResponse>(MsgID.LEAVE_PARTY_REQ,
    MsgID.LEAVE_PARTY_RESP, party.LeavePartyResponse.fromBuffer);
const kickPartyMember = MessageSpec<party.KickMemberResponse>(
    MsgID.KICK_PARTY_MEMBER_REQ,
    MsgID.KICK_PARTY_MEMBER_RESP,
    party.KickMemberResponse.fromBuffer);
const transferPartyLeader = MessageSpec<party.TransferLeaderResponse>(
    MsgID.TRANSFER_LEADER_REQ,
    MsgID.TRANSFER_LEADER_RESP,
    party.TransferLeaderResponse.fromBuffer);
const setPartyReady = MessageSpec<party.SetReadyResponse>(MsgID.SET_READY_REQ,
    MsgID.SET_READY_RESP, party.SetReadyResponse.fromBuffer);
const getMyParty = MessageSpec<party.GetMyPartyResponse>(MsgID.GET_MY_PARTY_REQ,
    MsgID.GET_MY_PARTY_RESP, party.GetMyPartyResponse.fromBuffer);

// Device plane (app_gateway WS 5201): registration / listing of the push
// targets for our account. app_gateway authenticates the session and pins
// user_id server-side; 6009 PUSH_NOTIFICATION is deliberately absent — the
// edge refuses it from clients.
const registerDevice = MessageSpec<notification.RegisterDeviceResponse>(
    MsgID.REGISTER_DEVICE_REQ,
    MsgID.REGISTER_DEVICE_RESP,
    notification.RegisterDeviceResponse.fromBuffer);
const unregisterDevice = MessageSpec<notification.UnregisterDeviceResponse>(
    MsgID.UNREGISTER_DEVICE_REQ,
    MsgID.UNREGISTER_DEVICE_RESP,
    notification.UnregisterDeviceResponse.fromBuffer);
const getUserDevices = MessageSpec<notification.GetUserDevicesResponse>(
    MsgID.GET_USER_DEVICES_REQ,
    MsgID.GET_USER_DEVICES_RESP,
    notification.GetUserDevicesResponse.fromBuffer);
